core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -360,41 +359,46 @@ class KafkaApis(val requestChannel: RequestChannel,
             requestChannel.noOperation(request.processor, request)
           }
         } else {
-          val response = ProducerResponse(produceRequest.correlationId,
-                                          mergedResponseStatus,
-                                          produceRequest.versionId,
-                                          delayTimeMs)
-          requestChannel.sendResponse(new RequestChannel.Response(request,
-                                                                  new RequestOrResponseSend(request.connectionId,
-                                                                                            response)))
+          val respHeader = new ResponseHeader(request.header.correlationId)
+          val respBody = if (request.header.apiVersion > 0)
```

Comment:
Should we be more explicit here wrt versioning? As is, when we add new versions we can accidentally get the wrong behavior if we miss updating this piece of code since the `request.header.apiVersion > 0` will catch all new versions. Explicitly handling each version and generating an error for the unexpected case seems safer (which should be impossible for now since it would be caught at parse time, but would catch the case where a new version is added but this code is not updated). Presumably this code will always need some updates with modifications to this request/response pair, but better safe than sorry.